### DISCUSSION TITLE
generic microdot websocket send and receive

### DISCUSSION THREAD

**Discussion by fabse-hack** [miguelgrinberg/microdot#197] [General]
Hello,

here is a snippet from my actualley micropython robot:

```python
app = Microdot()
Response.default_content_type = "html"

@app.route('/')
async def index(req):
    with open('robot.html', 'r') as f:
        html = f.read()
    ssid, ip_address, status = configure_wifi()
    return Response(html.replace('{{get_wifi_status}}', status)
                               .replace('{{get_ip_address}}', ip_address)
                               .replace('{{get_ssid}}', ssid))


@app.route("/ws")
@with_websocket
async def executeCarCommands(request, ws):

    while True:
        message = await ws.receive()
        websocket_message = json.loads(message)

        if "CMD" in websocket_message:
            direction = websocket_message["CMD"]
            if direction == 'forward':
                print('+forward')
                stop_auto()
                print('+auto_stopped')
                dc_motor.forward(100)
                display_text('FORWARD')
                response_message = {"message": "forward", "mode": "direct"}
                await ws.send(json.dumps(response_message))
            elif direction == 'back':
                print('+backwards')
                stop_auto()
                dc_motor.backwards(100)
                display_text('BACKWARD')
                response_message = {"message": "back", "mode": "direct"}
                await ws.send(json.dumps(response_message))
```

**can i change the code, that i can use ws.send and ws.receive generic in my main programm?**
for example here:

```python
@app.route('/')
async def index(req):
    with open('robot.html', 'r') as f:
        html = f.read()
    ssid, ip_address, status = configure_wifi()
    return Response(html.replace('{{get_wifi_status}}', status)
                               .replace('{{get_ip_address}}', ip_address)
                               .replace('{{get_ssid}}', ssid))
```

i will change this ".replace" to "ws.send".
I have already tried several modifications in the code, but it does not work as intended...

**Comment by miguelgrinberg** [comment]
Your index route is an HTTP route, it cannot receive or send WebSocket events. The HTML page that this route returns should make a WebSocket connection to a WebSocket route, and then you can send and receive from that second route.

**Reply by fabse-hack** [reply]
thank you for your replay. Puh, this is very hard for me, i'm still a beginner in python and the websocket route is not easy to understand. Can you provide me an example snippet?

**Reply by miguelgrinberg** [reply]
There is an example in this repository: https://github.com/miguelgrinberg/microdot/blob/main/examples/websocket/echo.py.

The index route returns an HTML page (index.html, located in the same directory). The HTML page has JavaScript code that makes a websocket connection to the WebSocket route.

**Reply by fabse-hack** [reply]
thanks, i will test it :)

**Reply by fabse-hack** [reply]
@miguelgrinberg this is not my usecase.

I want to send data from the backend (micropython, sensor data) to the html.
The sensor data are in different functions in the backend. 
So, the echo i can not use.

**Reply by miguelgrinberg** [reply]
I did not expect the echo example would work as is for you, it is just a demonstration. The point was to demonstrate how to organize the application to work with HTML and WebSocket routes. That is the same regardless of your use of sensors.

**Reply by fabse-hack** [reply]
> I did not expect the echo example would work as is for you, it is just a demonstration. The point was to demonstrate how to organize the application to work with HTML and WebSocket routes. That is the same regardless of your use of sensors.

oh okay, this only confused me.....
i still struggle on the code :(

**Reply by miguelgrinberg** [reply]
You initially asked about using `ws.send()` in your HTTP route, which cannot be done, because at the time the HTTP route is running the client does not have a WebSocket connection active. The response from the HTTP route will include (or reference) the JavaScript code that will make that WebSocket connection.

The echo example just demonstrates the separation between HTTP and WebSocket. If you need to use `ws.send` from a different function you can, but the client first needs to make the WebSocket connection, so you must have a WebSocket route similar to the echo one that receives the `ws` object. Once you have the object in the WebSocket route, you can pass it to other functions to use.

This is unfortunately not trivial to understand, you will need to try simpler examples and increase the complexity gradually. If you get the echo example to work, you can then maybe add sending a WebSocket message to the client when you press a button, or after some amount of time has passed.

**Reply by fabse-hack** [reply]
i'm still not have a solution.....

I'm try to connect to the websocket and send without dekorator.
Because in my ESP-Robot-Car, i have to use send() (from backend to frontend) is different places & functions....

and otherwise, from frontend - to backend in different places & functions.
I think this is to 80% complete wrong, but this is the way:

```python
from microdot import Microdot, Response, Request
from microdot.websocket import WebSocket
import asyncio
import json

app = Microdot()

@app.route('/')
async def index(request):
    html_content = """
    <!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Sensor Data</title>
            <script>
                const evtSource = new EventSource("http://127.0.0.1:5000/");
                evtSource.onmessage = function(event) {
                    const data = JSON.parse(event.data);
                    document.getElementById('ssid').textContent = "SSID: " + data.ssid;
                    document.getElementById('ip').textContent = "IP: " + data.ip;
                    document.getElementById('status').textContent = "Status: " + data.status;
                }
            </script>
    </head>
    <body>
        <h1>Sensor Data</h1>
        <p id="ssid">SSID: </p>
        <p id="ip">IP: </p>
        <p id="status">Status: </p>
    </body>
    </html>
    """
    response = Response(html_content)
    response.headers['Content-Type'] = 'text/html'
    return response

async def send_websocket_data():
    text = {"ip": "TESTTEST", "ssid": "dasdadad", "status": "dsdasdsdas"}
    json_data = json.dumps(text)
    headers = {"Host": "127.0.0.1:5000"}
    request = Request(app=app, client_addr=None, method='GET', url='/', http_version='1.1', headers=headers)
    web_socket_instance = WebSocket(request)
    await web_socket_instance.handshake()
    print("WebSocket connected successfully.")
    await web_socket_instance.send(json_data)

async def start_microdot_server():
    app.start_server(debug=True, port=5000)

async def main():
    task1 = asyncio.create_task(start_microdot_server())
    task3 = asyncio.create_task(send_websocket_data())
    await asyncio.gather(task1, task3)

if __name__ == '__main__':
    asyncio.run(main())
```

**Reply by miguelgrinberg** [reply]
Yes, there are many things that are wrong in your code, just to point the main two:

- Your client is connecting via server-sent events instead of WebSocket
- Your server does not have a WebSocket route

I really think you should start from the example I pointed you at, which does make a successful connection. In the WebSocket route of that example you can start your `send_websocket_data()` task. What you are doing now is invalid, you can't just create a `WebSocket` object from a made-up request, you need an actual connection from the client for this to work.

**Reply by fabse-hack** [reply]
hmm i'm still investigate your microdot. But i can not handle it correct.

1. Why the for i in range(100): and the while True: loop is not working?
2. Can you show me, how i can use "send" in other functions / places too with ws / sse ??? because when i implement ones, i can not use it again (i think, because of the dekorator)...


```python
from microdot import Microdot, Response
from microdot.websocket import with_websocket
from microdot.sse import with_sse
import json
import asyncio

app = Microdot()

async def configure_wifi():
    for i in range(100):
        ssid = str(i)
        ip_address = "12312121"
        status = "12312121212"
        asyncio.sleep(1)
        return ssid, ip_address, status

@app.route("/ws")
@with_websocket
async def get_sensor_data(request, ws):
    data = {"temperature": 25.5, "humidity": 50}
    await ws.send(json.dumps(data))

@app.route('/wifi')
@with_sse
async def wifi(request, sse):
    while True:
        async def generate():
            ssid, ip_address, status = await configure_wifi()
            data = {"ssid": ssid, "ip": ip_address, "status": status} 
            await sse.send(json.dumps(data))
        return await generate()

@app.route('/')
async def index(request):
    html_content = """
    <!DOCTYPE html>
    <html lang="en">
    <head>
        <meta charset="UTF-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>Sensor Data</title>
        <script>
            var socket = new WebSocket("ws://" + window.location.host + "/ws");

            socket.onmessage = function(event) {
                var sensorData = JSON.parse(event.data);
                document.getElementById("temperature").innerText = "Temperature: " + sensorData.temperature + " °C";
                document.getElementById("humidity").innerText = "Humidity: " + sensorData.humidity + " %";
            };

            const evtSource = new EventSource("/wifi");

            evtSource.onmessage = function(event) {
                const data = JSON.parse(event.data);
                document.getElementById('ssid').textContent = "SSID: " + data.ssid;
                document.getElementById('ip').textContent = "IP: " + data.ip;
                document.getElementById('status').textContent = "Status: " + data.status;
            };
            
        </script>
    </head>
    <body>
        <h1>Sensor Data</h1>
        <p id="temperature">Temperature: </p>
        <p id="humidity">Humidity: </p>
        <p id="ssid">SSID: </p>
        <p id="ip">IP: </p>
        <p id="status">Status: </p>
    </body>
    </html>
    """
    response = Response(html_content)
    response.headers['Content-Type'] = 'text/html'
    return response

if __name__ == '__main__':
    app.run(debug=True)
```

**Reply by miguelgrinberg** [reply]
Unfortunately I cannot explain code that I do not understand. Putting a `return` inside of a loop does not make much sense and you've done it on both loops.

**Reply by fabse-hack** [reply]
thanks!
I use the different app.routes, it's a workaround, basicly i want to have it generic.
this is working:
"yield" was new for me :D

```python
async def configure_wifi():
    for i in range(100):
        ssid = str(i)
        yield ssid
        await asyncio.sleep(1)


@app.route('/sensordata')
@with_websocket
async def sensordata(request, ws):
    async for ssid in configure_wifi():
        data = {"sensor": ssid} 
        await ws.send(json.dumps(data))

```

**Reply by miguelgrinberg** [reply]
This achieves the same result and is much simpler:

```python
@app.route('/sensordata')
@with_websocket
async def sensordata(request, ws):
    for ssid in range(100):
        data = {"sensor": str(ssid)} 
        await ws.send(json.dumps(data))
        await asyncio.sleep(1)
```

**Reply by fabse-hack** [reply]
thanks!
